### PR TITLE
[BTAT-2678] Completed ConfirmClientController and refactored Auth

### PR DESCRIPTION
**Pull request by MJCallahanPage**
**Note**
This PR includes the logic for adding the client VRN to the session cookie on submission of the `SelectClientVrnController.submit` method. 

On redirect to the `ConfirmClientVrnController.show` method a call is made to Auth to check for the existance of a delegated authority for the agent to act on behalf of the VRN held in session. 

If the authority does not exist it currently throws a FORBIDDEN response and renders the unauthorised view. Subsequent stories will update this to render the 'Your not authorised for this Client view'.

Subsequent tasks will also need to wipe the Client VRN from the Session Cookie.

The PR also includes component tests that were missing for `SelectClientVrnController` and `ConfirmClientVrnController`

Apologies for the size of the branch!

### CODE REVIEW

File: app/controllers/predicates/AuthPredicate.scala [hunk 1]
```diff
@@ -48,59 +47,41 @@ class AuthPredicate @Inject()(enrolmentsAuthService: EnrolmentsAuthService,
...
-      authenticateAsAgent.authorise(request, block)
+    if (enrolments.enrolments.exists(_.key == EnrolmentKeys.agentEnrolmentId)) {
+      Logger.debug("[AuthPredicate][checkAgentEnrolment] - Authenticating as agent")
+      authenticateAsAgentWithClient.invokeBlock(request, block)
     }
     else {
-      Logger.info(s"[AuthPredicate][checkAgentEnrolment] - Agent without HMRC-AS-AGENT enrolment. Enrolments: $enrolments")
-      Future(InternalServerError)
+      //TODO: Render Agent not Signed Up for Agent Services View
+      Logger.debug(s"[AuthPredicate][checkAgentEnrolment] - Agent without HMRC-AS-AGENT enrolment. Enrolments: $enrolments")
```
Comment (miranda-hawkes) [inline app/controllers/predicates/AuthPredicate.scala]: Do you need both these logger messages?

File: app/controllers/predicates/AuthoriseAsAgentOnly.scala [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+import uk.gov.hmrc.play.bootstrap.controller.FrontendController
+
+import scala.concurrent.Future
+
+@Singleton
+class AuthoriseAsAgentOnly @Inject()(enrolmentsAuthService: EnrolmentsAuthService,
+                                     val messagesApi: MessagesApi,
+                                     val serviceErrorHandler: ServiceErrorHandler,
+                                     implicit val appConfig: AppConfig
+                                    ) extends FrontendController with AuthBasePredicate with I18nSupport with ActionBuilder[Request] with ActionFunction[Request, Request] {
```
Comment (miranda-hawkes) [inline app/controllers/predicates/AuthoriseAsAgentOnly.scala]: sbt scalastyle is moaning about the length of this line

File: app/controllers/predicates/AuthoriseAsAgentOnly.scala [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  }
+
+  private def checkAgentEnrolment[A](enrolments: Enrolments, request: Request[A], f: Request[A] => Future[Result]) =
+    if (enrolments.enrolments.exists(_.key == EnrolmentKeys.agentEnrolmentId)) {
+      Logger.debug("[AuthoriseAsAgentOnly][checkAgentEnrolment] - Authenticated as agent")
+      f(request)
+    }
+    else {
+      //TODO: Render Agent not Signed Up for Agent Services View
+      Logger.debug(s"[AuthoriseAsAgentOnly][checkAgentEnrolment] - Agent without HMRC-AS-AGENT enrolment. Enrolments: $enrolments")
```
Comment (miranda-hawkes) [inline app/controllers/predicates/AuthoriseAsAgentOnly.scala]: Do we need both these loggers?

File: it/helpers/BaseIntegrationSpec.scala [hunk 4]
```diff
@@ -56,26 +64,35 @@ trait BaseIntegrationSpec extends WireMockHelper with GuiceOneServerPerSuite wit
...
-    def isAgentNotAuthenticated: PreconditionBuilder = {
-      AuthStub.unauthorisedNotLoggedIn()
+  class Agent()(implicit builder: PreconditionBuilder) {
+    def isAgentAuthorised: PreconditionBuilder = {
+      AuthStub.agentAuthorised()
       builder
     }
 
-    def isAgentNotEnrolled: PreconditionBuilder = {
+    def isAgentNotEnrolledToAsAgent: PreconditionBuilder = {
```
Comment (miranda-hawkes) [inline it/helpers/BaseIntegrationSpec.scala]: Could you rename this please to something a bit more self explanatory?

File: it/helpers/IntegrationTestConstants.scala [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  val VRN = "111111111"
+
+  val individualCustomerDetails = CustomerDetailsModel(
+    firstName = Some("Nanana"),
+    lastName = Some("VatMan"),
+    tradingName = Some("Vatmobile Taxi"),
+    organisationName = None
+  )
+
+  val organisationCustomerDetails = CustomerDetailsModel(
```
Comment (miranda-hawkes) [inline it/helpers/IntegrationTestConstants.scala]: Do we need organisation specific constants? This one isn't used either..

File: app/controllers/predicates/AuthPredicate.scala (hunk 1 above)
Comment (MJCallahanPage) [reply app/controllers/predicates/AuthPredicate.scala]: Yes, the one is for debugging and includes the enrolments and won't get output in production.

The warn gives us only the reason withou leaking potentially sensitive information and will be output in production for investigating issues in Kibana.

File: app/controllers/predicates/AuthoriseAsAgentOnly.scala (hunk 3 above)
Comment (MJCallahanPage) [reply app/controllers/predicates/AuthoriseAsAgentOnly.scala]: As per previous comment :)

**Pull request merged**